pkg/agent/controller/networkpolicy/reject.go

```diff
@@ -115,8 +117,27 @@ func (c *Controller) rejectRequest(pktIn *ofctrl.PacketIn) error {
 		isIPv6 = true
 	}
 
-	sIface, srcFound := c.ifaceStore.GetInterfaceByIP(srcIP)
-	dIface, dstFound := c.ifaceStore.GetInterfaceByIP(dstIP)
+	sIface, srcIsLocal := c.ifaceStore.GetInterfaceByIP(srcIP)
+	dIface, dstIsLocal := c.ifaceStore.GetInterfaceByIP(dstIP)
+
+	isFlexibleIPAMSrc, isFlexibleIPAMDst, ctZone, err := parseFlexibleIPAMStatus(pktIn, c.nodeConfig, srcIP, srcIsLocal, dstIP, dstIsLocal)
+	if err != nil {
+		return err
+	}
+	matches := pktIn.GetMatches()
+	if isFlexibleIPAMSrc && dstIsLocal || isFlexibleIPAMDst {
+		// For an ingress reject rule with AntreaIPAM cross-VLAN traffic, the original packet comes from uplink/gateway
+		// port, and the reject packet should be sent to uplink/gateway port, and the reject type should be
+		// RejectLocalToRemote.
+		// Overwrite dstIsLocal to false to correct the result from getRejectType.
+		if match := matches.GetMatchByName(binding.OxmFieldInPort); match != nil {
+			switch match.GetValue().(uint32) {
+			case c.nodeConfig.UplinkNetConfig.OFPort, c.nodeConfig.GatewayConfig.OFPort:
+				dstIsLocal = false
```

Comment:
Is it easier if we decide srcIsLocal/srcIsLocal by both the Interface type (containerInterface) and IP? Then we do not need to make such exceptions.

Reply:
This should be better but will have higher risk, I prefer to restrict the logic change for FleixbleIPAM cases only.
https://github.com/antrea-io/antrea/pull/4482#discussion_r1065397454

Reply:
@wenyingd : could you elaborate what you mean by "both the Interface type (containerInterface) and IP"?

@gran-vmv : I do not agree with the point about which approach has "higher risk". It is our code and we are not restricted by timeline either. We should make clear of all code logic and make a decision, not base on assumption of probability.

Reply:
@jianjuns This bug is coming from downstream but the root cause is on upstream, thus I hope we should fix the bug only, and file another PR for the refactor/enhancement if needed.

Reply:
Which upstream release you want the fix? In any case, I do not agree on a "quick" fix, unless you have a real timeline limit.

Reply:
Discussed with @edwardbadboy , we don't have a urgent timeline for this, and we can spend more time to enhance the quality. Thanks.

Reply:
For this question

> @wenyingd : could you elaborate what you mean by "both the Interface type (containerInterface) and IP"?

Currently we decide srcIsLocal/srcIsLocal by IP only, and the result is not accurate in cross-VLAN traffic as we patched in this PR. Thus we can consider to decide srcIsLocal/srcIsLocal by IP and interface.

Reply:
Yes, my thought is we could firstly get InterfaceConfig by the IP address, and then we can check the type of InterfaceConfig. We return true for srcIsLocal/dstIsLocal only when the returned InterfaceConfig type is ContainerInterface. In a result, we don't need to think about that case that the interface is for gateway or uplink. Further more, we could check the hostInterface type in VM Agent case ( for there is no container interface in that scenario) to decide if it is local. In this way, I think the code is more unified and clean

Reply:
Refactored the logic for `isLocal`.